### ISSUE TITLE
NPE evaluating JS eval() in double-nested IILE from Java/Kotlin

### ISSUE THREAD

**Issue by Tradias** [oracle/graal#3029]
**Describe GraalVM and your environment :**
 - GraalVM version or commit id if built from source: **20.3.0**
 - CE or EE: **CE**
 - JDK version: **JDK15**
 - OS and OS Version: **Windows 10 Pro**
 - Architecture: **amd64**
 - The output of `java -Xinternalversion`:
```
OpenJDK 64-Bit Server VM (15+36-1562) for windows-amd64 JRE (15+36-1562), built on Aug 11 2020 22:33:04 by "mach5one" with unknown MS VC++:1925
```

**Have you verified this issue still happens when using the latest snapshot?**
no

**Describe the issue**
Evaluating the JS script `(() => {  (() => {  eval()  })()  })()` from Java/Kotlin causes NPE in GraalJSTranslator.createThisNode.

**Code snippet or code repository that reproduces the issue**
```java
import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Source;

public class App {
    public static void main(String[] args) {
        var script = "(() => {  (() => {  eval()  })()  })()";
        try (var it = Context.newBuilder().build()) {
            it.eval(Source.create("js", script));
        }
    }
}
```

**Steps to reproduce the issue**
Assuming the above code snippet is in a file called `App.java`:
1. `javac.exe -cp graal-sdk-20.3.0.jar app.java`
2. `java.exe -cp "graal-sdk-20.3.0.jar;truffle-api-20.3.0.jar;js-20.3.0.jar;icu4j-67.1.jar;." App`

**Expected behavior**
The code runs without throwing an exception.

**Additional context**
Stacktrace:

```
Exception in thread "main" org.graalvm.polyglot.PolyglotException: java.lang.NullPointerException: Cannot invoke "com.oracle.truffle.js.parser.env.Environment$VarRef.createReadNode()" because the return value of "com.oracle.truffle.js.parser.env.Environment.findThisVar()" is null
	at com.oracle.truffle.js.parser.GraalJSTranslator.createThisNode(GraalJSTranslator.java:1505)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterFunctionNode(GraalJSTranslator.java:412)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterFunctionNode(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterFunctionNode(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.FunctionNode.accept(FunctionNode.java:331)
	at com.oracle.js.parser.ir.LexicalContextNode.accept(LexicalContextNode.java:85)
	at com.oracle.js.parser.ir.LexicalContextExpression.accept(LexicalContextExpression.java:68)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterCallNode(GraalJSTranslator.java:2192)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterCallNode(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterCallNode(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.CallNode.accept(CallNode.java:154)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterExpressionStatement(GraalJSTranslator.java:3247)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterExpressionStatement(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterExpressionStatement(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.ExpressionStatement.accept(ExpressionStatement.java:85)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transformStatementInBlock(GraalJSTranslator.java:1416)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transformStatements(GraalJSTranslator.java:1329)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterBlock(GraalJSTranslator.java:1221)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterBlock(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterBlock(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.Block.accept(Block.java:174)
	at com.oracle.js.parser.ir.LexicalContextNode.accept(LexicalContextNode.java:85)
	at com.oracle.js.parser.ir.Block.accept(Block.java:360)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.translateFunctionBody(GraalJSTranslator.java:428)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterFunctionNode(GraalJSTranslator.java:393)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterFunctionNode(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterFunctionNode(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.FunctionNode.accept(FunctionNode.java:331)
	at com.oracle.js.parser.ir.LexicalContextNode.accept(LexicalContextNode.java:85)
	at com.oracle.js.parser.ir.LexicalContextExpression.accept(LexicalContextExpression.java:68)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterCallNode(GraalJSTranslator.java:2192)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterCallNode(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterCallNode(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.CallNode.accept(CallNode.java:154)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterExpressionStatement(GraalJSTranslator.java:3247)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterExpressionStatement(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterExpressionStatement(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.ExpressionStatement.accept(ExpressionStatement.java:85)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transformStatementInBlock(GraalJSTranslator.java:1416)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transformStatements(GraalJSTranslator.java:1329)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterBlock(GraalJSTranslator.java:1221)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterBlock(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterBlock(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.Block.accept(Block.java:174)
	at com.oracle.js.parser.ir.LexicalContextNode.accept(LexicalContextNode.java:85)
	at com.oracle.js.parser.ir.Block.accept(Block.java:360)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.translateFunctionBody(GraalJSTranslator.java:428)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterFunctionNode(GraalJSTranslator.java:393)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.enterFunctionNode(JavaScriptTranslator.java:58)
	at com.oracle.truffle.js.parser.GraalJSTranslator.enterFunctionNode(GraalJSTranslator.java:182)
	at com.oracle.js.parser.ir.FunctionNode.accept(FunctionNode.java:331)
	at com.oracle.js.parser.ir.LexicalContextNode.accept(LexicalContextNode.java:85)
	at com.oracle.js.parser.ir.LexicalContextExpression.accept(LexicalContextExpression.java:68)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transform(GraalJSTranslator.java:218)
	at com.oracle.truffle.js.parser.GraalJSTranslator.transformFunction(GraalJSTranslator.java:292)
	at com.oracle.truffle.js.parser.GraalJSTranslator.translateScript(GraalJSTranslator.java:287)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.translateScript(JavaScriptTranslator.java:99)
	at com.oracle.truffle.js.parser.JavaScriptTranslator.translateScript(JavaScriptTranslator.java:74)
	at com.oracle.truffle.js.parser.GraalJSEvaluator.parseScript(GraalJSEvaluator.java:265)
	at com.oracle.truffle.js.lang.JavaScriptLanguage.parseScript(JavaScriptLanguage.java:278)
	at com.oracle.truffle.js.lang.JavaScriptLanguage.parse(JavaScriptLanguage.java:187)
	at com.oracle.truffle.api.TruffleLanguage$ParsingRequest.parse(TruffleLanguage.java:848)
	at com.oracle.truffle.api.TruffleLanguage.parse(TruffleLanguage.java:1495)
	at com.oracle.truffle.api.LanguageAccessor$LanguageImpl.parse(LanguageAccessor.java:311)
	at com.oracle.truffle.polyglot.PolyglotSourceCache.parseImpl(PolyglotSourceCache.java:94)
	at com.oracle.truffle.polyglot.PolyglotSourceCache.access$300(PolyglotSourceCache.java:56)
	at com.oracle.truffle.polyglot.PolyglotSourceCache$WeakCache.lookup(PolyglotSourceCache.java:223)
	at com.oracle.truffle.polyglot.PolyglotSourceCache.parseCached(PolyglotSourceCache.java:80)
	at com.oracle.truffle.polyglot.PolyglotLanguageContext.parseCached(PolyglotLanguageContext.java:369)
	at com.oracle.truffle.polyglot.PolyglotContextImpl.eval(PolyglotContextImpl.java:940)
	at org.graalvm.polyglot.Context.eval(Context.java:347)
	at evaljs.Application.main(Application.kt:12)
Caused by: Attached Guest Language Frames (0)
Internal GraalVM error, please report at https://github.com/oracle/graal/issues/.
```
</details>

**Comment by mcraj017** [comment]
@Tradias Thanks for reporting the bug
We will check it and let you know the findings

**Comment by mcraj017** [comment]
@Tradias  Please share the command line used to run your code snippet

**Comment by Tradias** [comment]
I have updated the issue description. Also changed the example to Java, I guess that makes it easier for you to reproduce

**Comment by mcraj017** [comment]
@Tradias thank you

**Comment by mcraj017** [comment]
@Tradias This has been added to our feature list and will be fixed in future versions

**Comment by monsters4r** [comment]
I'm getting this same error - but in a much more complicated scenario.  I'm trying to call a typescript code segment that essentially uses the @antv/G2 library - the code is just a simple example that is based on the example on this page - https://g2.antv.vision/en/docs/manual/about-g2.  

I took this polyglot example https://cwiki.apache.org/confluence/display/NETBEANS/Apache+NetBeans+extension+for+Visual+Studio+Code, made sure it worked, then substituted in the G2 example.  

Because I needed to "import" the G2 library (as seen in the G2 example), I used webpack to create a javascript file that avoided the use of import.  

This issue is the only reference I can find to the same error I'm seeing.    I'm attaching a stack trace.
[stacktrace.txt](https://github.com/oracle/graal/files/5837718/stacktrace.txt)

Any thoughts are appreciated.
